Add blog post about Rails upgrade (#110)
simplificator/simplificator-blog: content/posts/rails-upgrade-strategy.md

One more example. When I once tried to do a Rails upgrade without updating dependencies in advance, all tests were broken. The error message was very weird and not helpful at all. Google and Stackoverflow also did not know about such a problem. I changed my approach and started to update gems one by one. It turned out that the `database-cleaner` gem, that our application uses, introduced some breaking changes since our last update (it actually became a set of gems depending on the used database adapters). In our case I had to install a new gem `database-cleaner_mongoid` and do some configuration changes. The problem was solved!

Another option to perform Rails upgrades smoothly is to keep your dependencies up to date. You can do it easily by installing [dependabot](https://docs.github.com/en/code-security/dependabot). It will run your pipeline against any new gem updates and you will be able to catch breaking changes early.

> Often before the Rails upgrade you would also like to upgrade to a newer Ruby version. Here the same principle is applied: if the upgrade does not work out of the box and you get errors, try to update first those dependencies that block the Ruby upgrade.

Author: Thanks @pinefoambath ! It's a very good point why to do an upgrade at all, I added this paragraph to my text. But I am not sure about proposing to perform Rails upgrades as one of our services. I thought the main target group of this post would be developers. But maybe it still makes sense to advertise such a service? What do you think @crackofdusk ?

---
title: "How-To: Rails upgrade strategy"

> date: 2022-05-25

Reviewer: Don't forget to update the publication date before merging.

It might seem like a lot of work because the list of dependencies can be pretty long (I ended up with one pull request updating 20 gems plus 4 pull requests for individual gem updates). But at the end of the day it saves you much time since you can discover faster which dependency caused the errors after its update.

Another option to perform Rails upgrades smoothly is to keep your dependencies up to date. You can do it easily by installing [dependabot](https://docs.github.com/en/code-security/dependabot). It will run your pipeline against any new gem updates and you will be able to catch breaking changes early.

Often before the Rails upgrade you would also like to upgrade to a newer Ruby version. Here the same principle is applied: if the upgrade does not work out of the box and you get errors, try to update first those dependencies that block the Ruby upgrade.

Here once again all steps recapped:

1. Upgrade Ruby version to the latest compatible with the desired Rails version
2. If this does not work, individually upgrade gems blocking Ruby update, and try again

> 3. Update Rails to the next patch version

Reviewer: ```suggestion
3. Update Rails to the most recent patch version of the currently used release series
```

> 4. Run `bundle outdated` and update one by one all outdated gems

Reviewer: ```suggestion
4. Run `bundle outdated` and update all outdated gems one by one
```

> 5. Perform Rails upgrade to a minor or a major version

Reviewer: ```suggestion
5. Perform the Rails upgrade to a minor or a major version following the [Rails upgrade guide](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html)
```
